oauthenticator/globus.py

```diff
@@ -109,24 +119,42 @@ def _scope_default(self):
             'urn:globus:auth:scope:transfer.api.globus.org:all'
         ]
 
-    globus_local_endpoint = Unicode(help="""If Jupyterhub is also a Globus endpoint, its endpoint id can
-                be specified here.""").tag(config=True)
+    allow_refresh_tokens = Bool(
+        help="""Allow users to have Refresh Tokens. If Refresh Tokens are not
+        allowed, users must use regular Access Tokens which will expire after
+        a set time. Set to False for increased security, True for increased
+        convenience."""
+    ).tag(config=True)
+
+    def _allow_refresh_tokens_default(self):
+        return True
+
+    globus_local_endpoint = Unicode(help="""If Jupyterhub is also a Globus
+    endpoint, its endpoint id can be specified here.""").tag(config=True)
 
     def _globus_local_endpoint_default(self):
         return os.getenv('GLOBUS_LOCAL_ENDPOINT', '')
 
     def pre_spawn_start(self, user, spawner):
-        """Add tokens to the spawner whenever the spawner starts a notebook. Token names are translated to conform
-        to bash-style variable names. For example, 'transfer.api.globus.org' becomes 'TRANSFER_API_GLOBUS_ORG'"""
-        env_tokens = {
-            self.bashify_variable_name(name): token for name, token in user.auth_state['transfer_tokens'].items()
-        }
+        """Add tokens to the spawner whenever the spawner starts a notebook.
+        This will allow users to create a transfer client:
+        globus-sdk-python.readthedocs.io/en/stable/tutorial/#tutorial-step4
+        """
         if self.globus_local_endpoint:
-            spawner.environment.update({'GLOBUS_LOCAL_ENDPOINT': self.globus_local_endpoint})
-        spawner.environment.update(env_tokens)
-
-    def bashify_variable_name(self, var_name):
-        return var_name.replace('.', '_').replace('-', '_').upper()
+            spawner.environment.update(
+                {'GLOBUS_LOCAL_ENDPOINT': self.globus_local_endpoint}
+            )
+        if user.auth_state.get('globus_data'):
+            globus_data = base64.b64encode(
+                pickle.dumps(user.auth_state['globus_data'])
+            )
+            spawner.environment['GLOBUS_DATA'] = globus_data
```

Comment:
Let's walk through an example. As stated before, this will change and become transparent once we have a shim layer setup, but these are what the manual steps will look like.

First off, a user logs into Jupyterhub. They need to get their tokens:

    import pickle, base64, os
    data = pickle.loads(base64.b64decode(os.getenv('GLOBUS_DATA')))

`data` will look like:

    {
        'client_id': '4efe3069-e748-4b4d-8c5f-fbe515dc5e42', 
        'tokens': {
            'transfer.api.globus.org': {
                'refresh_token': <RefreshToken>, 
                'access_token': <AccessToken>,
                <More Junk>
            },
            ...<More Tokens here>...
        }
    }

For most purposes, only the access token will be needed, and the result looks like this:

    import globus_sdk
    transfer_authorizer = globus_sdk.AccessTokenAuthorizer(data['tokens']['transfer.api.globus.org']['access_token'])
    tc = globus_sdk.TransferClient(authorizer=transfer_authorizer)

`tc` is everything the user should need. They can check their endpoints, do transfers, check tasks, etc. The problem is `tc` will expire after a while.

If the user wants to do a long running task, and they really want to use Refresh Tokens, they need a better Authorizer. To my knowledge, `AccessTokenAuthorizer` isn't capable of handling refresh tokens. We need a better class capable of doing the 'refreshing'. Although NativeApp is typically used for deriving tokens, it can also do the refreshing we require. The users code then looks like this:

    client = globus_sdk.NativeAppAuthClient(data['client_id'])
    transfer = data['tokens']['transfer.api.globus.org']

    authorizer = globus_sdk.RefreshTokenAuthorizer(transfer['refresh_token'], client, access_token=transfer['access_token'], expires_at=transfer['expires_at_seconds'])
    tc = globus_sdk.TransferClient(authorizer=authorizer)

Everything else is the same as before, but now the user can use `tc` until the end of time.

To my knowledge, `AccessTokenAuthorizer` can't handle refresh tokens. If it could, I would say we should use that instead. If you know a better way to get a Transfer Client using Refresh Tokens, I think that would be a better option.